File: server/mc/1.12.2-cleanroom/src/main/java/calebxzhou/rdi/mc/server/rcmd/RcmdCommandSenderSource112.java
package calebxzhou.rdi.mc.server.rcmd;

import calebxzhou.rdi.mc.rcmd.RcmdSource;
import net.minecraft.command.ICommandSender;
import net.minecraft.entity.Entity;
import net.minecraft.entity.player.EntityPlayerMP;
import net.minecraft.util.text.TextComponentString;

import java.util.UUID;

public final class RcmdCommandSenderSource112 implements RcmdSource {
    private final ICommandSender sender;

    public RcmdCommandSenderSource112(ICommandSender sender) {
        this.sender = sender;
    }

    public EntityPlayerMP getPlayer() {
        Entity entity = sender.getCommandSenderEntity();
        return entity instanceof EntityPlayerMP player ? player : null;
    }

    @Override
    public String name() {
        return sender.getName();
    }

    @Override
    public UUID playerId() {
        EntityPlayerMP player = getPlayer();
        return player == null ? RcmdSource.NO_PLAYER_ID : player.getUniqueID();
    }

    @Override
    public boolean hasPermission(String permission) {
        return sender.canUseCommand(4, permission);
    }

    @Override
    public void sendFeedback(String message) {
        sender.sendMessage(new TextComponentString(message));
    }

    @Override
    public void sendError(String message) {
        sender.sendMessage(new TextComponentString("[rcmd] " + message));
    }
}

File: server/mc/1.12.2-cleanroom/src/main/kotlin/calebxzhou/rdi/mc/server/rcmd/RcmdCommandSenderSource112.kt
package calebxzhou.rdi.mc.server.rcmd

import calebxzhou.rdi.mc.rcmd.RcmdSource
import net.minecraft.command.ICommandSender
import net.minecraft.entity.player.EntityPlayerMP
import net.minecraft.util.text.TextComponentString
import java.util.UUID

class RcmdCommandSenderSource112(private val sender: ICommandSender) : RcmdSource {
    val player: EntityPlayerMP?
        get() = sender.commandSenderEntity as? EntityPlayerMP

    override fun name(): String = sender.name

    override fun playerId(): UUID = player?.uniqueID ?: RcmdSource.NO_PLAYER_ID

    override fun hasPermission(permission: String): Boolean = sender.canUseCommand(4, permission)

    override fun sendFeedback(message: String) {
        sender.sendMessage(TextComponentString(message))
    }

    override fun sendError(message: String) {
        sender.sendMessage(TextComponentString("[rcmd] $message"))
    }
}
